Espnet3/install only

Masao-Someki opened the pull request:
## What?

Update installation of espnet. With this PR, we can install espnet in a structured way.
E.g., `pip install espnet[all]`, `pip install espnet[core,task-asr]`

## Why?

The dependencies of espnet becomes very large. We wanted to split it.

## See also

#6133
#6135


## Note

This PR might include formatting automatically done by precommit.ci, but that will be merged to the master branch in another PR. This PR should only be merged after that PR.

Masao-Someki commented:
@Fhrozen just commented on my that the devcontainer-related script should be the another PR, as it is not specifically related on installation.

I will remove those, and after merging the #6140 the number of files should be 4 (pyproject.toml, Makefile, setup.py, and .gitignore)

sw005320 commented:
Please fix all CI issues.

Fhrozen commented on .github/workflows/ci_on_ubuntu.yml:
```diff
@@ -316,6 +316,7 @@ jobs:
         python3 -m pip install -U numba
         ./tools/installers/install_torch.sh false ${TH_VERSION} CPU
         ./tools/installers/install_chainer.sh CPU
+        cp ./tools/setup.py ./ && mv pyproject.toml pyproject.toml.tmp # && mv espnet3 .espnet3
```
I did this to keep the dependencies of ESPnet 1/2 and ESPnet 3 separate. Because espnet1/2 are installed by the make file, I make the  `setup.py` to move between folders. I looked for keeping both files `pyproject.toml` and `setup.py` in the same directory and redirecting the installation when the `espnet3` is added, but it was a little messy. Let me know if this option is not valid, and will try to look for a different solution.

sw005320 replied:
Would it be solved in the future?
If so, we can keep it.
Since most of our espnet2 integration test is working, I think we can unify it with `pyproject.toml`

Masao-Someki commented on pyproject.toml:
```diff
@@ -0,0 +1,139 @@
+  "h5py",
+  "lhotse",
+  "jamo",
+  "torch_complex",
```
I found that if we want to move torch_complex to task-specific installation, we need to mofdify 63 additional files. This is mainly used in asr/diar/enh and many test scripts, I'm thinking of moving it to `core` package..

Masao-Someki commented:
Oh, I just realized that for this PR the formatting is applied more than #6140...? I will check why

Masao-Someki commented:
I found that if I run black in local PC it generates different output from precomit's output. It seems that the current master is formatted on local PC. I will fix that.

Masao-Someki commented:
Hi, here's a description and discussion items on how I changed each packages:

* **pyworld**: It's used in almost all the `evaluate_f0` and `evaluate_uvu` scripts under each task. Judging by the script names, it seems like it's mainly used in TTS/VC. Should we keep it as a core dependency, or is it okay to move to task-tts?

* **lightning**: It's mainly used in ESPnet-3, and although it's supported in ESPnet-2 as a LightningTrainer, it feels a bit off to include it as a core dependency of ESPnet-2.. I'd like to have comments on this package.

* **nltk**: It's only used for evaluation in MT, ST, and dialogue. I'm not sure which dependency group is reasonable.

* **h5py**: Other than being used in `ESPnetDataset`, it's only used in tests and in espnet1. I moved it to the test dependency group, and making the dataset import it only when the h5py dataset type is actually used.

* **lhotse**: This is mainly used in diar-ASR/ASR and a few data preparation scripts under `egs2/`. It's not used at all in the main ESPnet code, so I removed it from the dependencies.

* **jamo**: This is used for Korean G2P. Moved it under TTS.

* **protobuf**: I couldn't find any actual usage under `egs/`, `egs2/`, `espnet/`, or `espnet2/`, so I removed it.

* **espnet_model_zoo**: Should we include this package in somewhere?

* **task-asr-state-space**: There are several packages that is only required from state-space model definition. so I moved them to `task-asr-state-space`.

* **tensorboard**, **torch_complex**, **einsum**, : I moved it to core, as it is used in several tasks.



Let me know your thoughts on each of these.

Masao-Someki commented:
I've created a new venv, run `pip install -e .[core]`, and ASR inference with the following model runs without issue.

 https://huggingface.co/pyf98/librispeech_100_ctc_e_branchformer

Masao-Someki commented:
I'm not sure why, but I'm getting different formatting results from Black compared to what @Fhrozen gets on his local machine, even if we are using the same version. This inconsistency is the main reason we're seeing so many formatting changes in the files.

I asked him to run the formatter on his end to resolve the differences, which should make the changes disappear, but I don't think this is a sustainable situation...

Masao-Someki commented:
Not sure why the setuptools version has changed, but we need to use specific version to avoid import error of chainer. 
I found that in the CI the setuptools version is 79.0.1, but we need setuptools<74.0.0

Reference:
![image](https://github.com/user-attachments/assets/ede55fed-d333-45d7-966f-618f989f723c)

sw005320 commented on doc/installation.md:
````diff
@@ -173,8 +173,81 @@ We also have [prebuilt Kaldi binaries](https://github.com/espnet/espnet/blob/mas
+
+    ```sh
+    $ cd <espnet-root>
+    $ pip install -e ".[core,task-asr]"
````
1. Is it possible to automatically include core when we specify task-*?
2. what is the default behavior without specifying it.
3. I still cannot understand the role of `core`. If it is used as a standalone? If not, I cannot understand why we make it as a package.

sw005320 commented on ci/test_integration_espnet2.sh:
```diff
@@ -148,7 +157,14 @@ done
 rm -rf exp dump data
 cd "${cwd}"
 
+# Uninstall task-dependency
+python3 test_utils/uninstall_extra.py
```
Can this uninstall `[task-asr]` libraries?
Why not use `pip uninstall`?

Masao-Someki replied:
I tried `pip uninstall espnet[task-asr]` but it will uninstall only the espnet, and not the packages in [task-asr]. I used several commands/tools to uninstall only [task-asr] related packages, and this is the best way so far. Other tools couldn't uninstall dependencies or agrgessively uninstall everything including pip.

![image](https://github.com/user-attachments/assets/98358961-94d0-4666-b48b-06357a747593)
![image](https://github.com/user-attachments/assets/763406ea-3b92-4fe8-b06d-84fc993b1c8e)
![image](https://github.com/user-attachments/assets/0595ef95-3fae-4d0c-a052-334884705a59)

sw005320 commented on doc/installation.md:
````diff
@@ -173,8 +173,80 @@ We also have [prebuilt Kaldi binaries](https://github.com/espnet/espnet/blob/mas
+1. Install ESPnet via `pyproject.toml`
+
+    ESPnet adopts the modern Python packaging standard using `pyproject.toml`.
+    You can install the **core dependencies** with:
+
+    * Core installation:
+
+        ```sh
+        $ cd <espnet-root>
+        $ pip install -e ".[core]"
+        ```
````
We should remove the descriptions about `core`

sw005320 commented on doc/installation.md:
```diff
@@ -173,8 +173,80 @@ We also have [prebuilt Kaldi binaries](https://github.com/espnet/espnet/blob/mas
+
+        | Group      | Purpose                       |
+        | ---------- | ----------------------------- |
+        | `core`     | Core libraries and runtime    |
```
We should remove

Masao-Someki commented:
@Fhrozen Once the CI issues are resolved, could you run the formatter on this branch? There have been several updates to files in `espnet2/` to avoid import errors, but the formatting differences make it hard to identify the actual changes.

sw005320 commented on espnet2/asr/state_spaces/utils.py:
```diff
@@ -4,8 +4,15 @@
 
-import hydra
-from omegaconf import DictConfig, ListConfig
+try:
+    import hydra
+    from omegaconf import DictConfig, ListConfig
+except ImportError:
+    import logging
+
+    logging.warning(
+        "If you are running state-space model," "run `pip install espnet['task-asr']`."
+    )
```
Is it correct?
`hydra` and `omegaconf` are related to `espnet-3`.
Why do they appear here?
What we need to install `task-asr`

Masao-Someki replied:
It looks like the files under `espnet2/asr/state_spaces/` use Hydra and OmegaConf.
Also, `opt_einsum` is only used within this directory, so I’ve included these in the `task-asr` group.

sw005320 commented on pyproject.toml:
```diff
@@ -0,0 +1,162 @@
+
+[project.optional-dependencies]
+
+espnet-3 = [
```
I cannot understand the position of this.
How is it used?

Masao-Someki replied:
It will be used in espnet-3, but maybe we should add this dependency in future PR.

sw005320 commented:
@Masao-Someki and @Fhrozen, this might help



To reflect flake8 in the `.pre-commit-config.yaml` file in the `@espnet/espnet` repository, you should add a flake8 hook configuration under the `repos:` section. This ensures that flake8 will run as part of your pre-commit checks.

Here’s how you can do it:

1. **Locate `.pre-commit-config.yaml`**  
   Open the `.pre-commit-config.yaml` file at the root of your repository (or in the `/files` directory if that’s where it is).

2. **Add or Update the flake8 Hook**  
   Add the following block under the `repos:` section if it doesn’t exist, or update it if it does:

   ```yaml
   - repo: https://github.com/pycqa/flake8
     rev: 6.1.0  # or the latest stable version
     hooks:
       - id: flake8
         additional_dependencies: []  # Add any flake8 plugins you use here
   ```

3. **Example of a Complete Section:**

   ```yaml
   repos:
     - repo: https://github.com/pycqa/flake8
       rev: 6.1.0
       hooks:
         - id: flake8
           additional_dependencies: [flake8-bugbear]
           # You can add more flake8 plugins to the array if needed
   ```

4. **Install/Update Pre-commit Hooks**  
   After editing the file, run:
   ```
   pre-commit install
   pre-commit autoupdate
   ```

**Summary:**  
- Add the flake8 repo block to your `.pre-commit-config.yaml`
- Specify the `rev` (version)
- Use the `additional_dependencies` list for any plugins

Would you like to see the current contents of your `.pre-commit-config.yaml` file to get a context-aware example?

Fhrozen commented on ci/test_integration_espnetez.sh:
```diff
@@ -17,6 +17,14 @@ gen_dummy_coverage(){
+# First uninstall all espnet-related dependencies including all extras.
+# I use toml and load the pyproject.toml
+python3 -m pip install toml
+python3 test_utils/uninstall_extra.py
```
this line is uninstalling nltk, because it is an optional package. Check if the package should become a permanent package or remove the test.

Fhrozen replied:
Or maybe disable the test related to the package.

Fhrozen replied:
nvm, I just found the solution

Masao-Someki replied:
@Fhrozen Sonce NOTK is now used in many tasks, I think we can simply move it to the main dependency?

Masao-Someki replied:
Also it will fix the CI issue?

sw005320 commented on tools/Makefile:
```diff
@@ -128,7 +128,9 @@ lightning.done: pytorch.done
 # NOTE(kamo): conda_packages is not necessary for installation of espnet, but add it the dependencies just in case.
 espnet.done: pytorch.done conda_packages.done lightning.done
 	@echo NUMPY_VERSION=$(shell . ./activate_python.sh && python3 -c "import numpy; print(numpy.__version__)")
```
I believe this is not resolved.
Why do we still maintain `setup.py` and `pyproject.toml`?

In my understanding, the current `all` in `pyproject.toml` is almost the same as `setup.py`
(Thus, the integration tests are all passed.)
Can you check them and resolve the confusing situations?

Fhrozen commented on pyproject.toml:
```diff
@@ -0,0 +1,157 @@
+  "sphinx-markdown-builder"
+]
+
+all = [
+  "espnet[task-asr]",
+  "espnet[task-tts]",
+  "espnet[task-enh]",
+  "espnet[task-spk]",
+]
```
From Gemini:

That is an excellent and insightful question that explores the boundaries of what pyproject.toml is designed to do.

The short answer is **no**, it is not possible to use wildcards (`task-*`) or loops directly within the `pyproject.toml` file to dynamically group dependencies.

### The Reason: pyproject.toml is Declarative, Not Programmatic
The `pyproject.toml file is designed to be a declarative and static configuration file. This means you declare what your project's configuration is, but you cannot include logic—like loops, conditionals, or wildcards—to compute that configuration. Its strength lies in its simplicity and predictability; any tool can read the file and know the project's metadata without executing any code.

### The Standard Way: Explicitly Listing the Extras
The correct and standard way to achieve your goal is to explicitly list each extra you want to include in your my-packages group. While it may feel repetitive, it is unambiguous and the officially supported method.

```toml
# pyproject.toml

[build-system]
requires = ["setuptools>=61.0"]
build-backend = "setuptools.build_meta"

[project]
name = "my-component-package"
version = "1.0.0"

# 1. Define your individual component extras
[project.optional-dependencies]
task-1 = [
    "pandas",
]
task-2 = [
    "matplotlib",
    "seaborn"
]
task-3 = [
    "scikit-learn"
]
# ...and so on for task-4, task-5, etc.

# 2. Create the group that explicitly includes the others
all-tasks = [
    "my-component-package[task-1]",
    "my-component-package[task-2]",
    "my-component-package[task-3]",
]

# You might also want a group for development tools
dev = [
    "my-component-package[all-tasks]", # Include all the above
    "pytest",
    "ruff"
]
```

### The Advanced Way: Adding Dynamic Logic with `setup.py`
If you absolutely must have dynamic logic (for example, in a very complex project with dozens of auto-generated extras), you can fall back to using a `setup.py` script.

This involves telling `pyproject.toml` that the `optional-dependencies` are "dynamic," meaning they will be determined at build time by the build backend (`setuptools`) running the `setup.py` script.

**This adds complexity and is generally not recommended unless you have a strong reason.**

### Conclusion

| Method|	How It Works|	Pros	|Cons|
|:--:|:--:|:--:|:--:|
|Standard (`pyproject.toml` only)	|Explicitly list each extra in the group.	|Simple, clear, standard, easy for tools to analyze.	|Requires manual updating if you add a new `task-N`.|
|Advanced (Dynamic `setup.py`)	|Use Python code to generate the list of extras.	|Automated, powerful for complex scenarios.	|Adds complexity, dependencies are not statically visible, can be harder to debug.|

Masao-Someki replied:
@Fhrozen Could you try copilot review? If it can detect this error, we can continue with that.

Masao-Someki commented:
Okay, it seems we need additional test that checks `all` includes all dependencies. I'll prepare that in `test_import_all.py`

sw005320 commented on tools/Makefile:
```diff
@@ -128,7 +128,9 @@ lightning.done: pytorch.done
 # NOTE(kamo): conda_packages is not necessary for installation of espnet, but add it the dependencies just in case.
 espnet.done: pytorch.done conda_packages.done lightning.done
 	@echo NUMPY_VERSION=$(shell . ./activate_python.sh && python3 -c "import numpy; print(numpy.__version__)")
+	cp ./setup.py ../ && mv ../pyproject.toml ../pyproject.toml.tmp
```
This must be changed.

sw005320 commented on pyproject.toml:
```diff
@@ -21,22 +21,37 @@ classifiers = [
 requires-python = ">=3.10"
 license = { "text" = "Apache Software License" }
 dependencies = [
```
how about using `==` for stable installation?

Masao-Someki replied:
Thank you, this is a good idea, especially for packages like formatting tools or other libraries that don’t get frequent updates. 
However, there are some libraries, like `datasets`, that are very actively developed (new version in almost one month). For these, using the specific version might prevent us from benefiting from important bug fixes or new features in a timely manner.

So how about using `==` for more stable or rarely updated packages, and using a more flexible version specifier (like `>=`) for actively maintained ones?

https://pypi.org/project/datasets/#history

sw005320 commented on pyproject.toml:
```diff
@@ -20,38 +20,40 @@ classifiers = [
 requires-python = ">=3.10"
 license = { "text" = "Apache Software License" }
+
+# We can use newer setuptools after removing chainer support.
```
It's better to make a comment for some packages that strictly require a specific version, e.g., `"sentencepiece==0.2.0",` (currently == or <=).

Also, what is the process of upgrading them?
When we fix the version, we can test upgrading to it. If it is successful, we will replace the versions with the latest ones.

Masao-Someki replied:
I think we can use Dependabot or GitHub Actions to automatically comment or create PR.

- Dependabot: Automatically checks `pyproject.toml` and creates PRs for outdated packages.

- GitHub Actions + Scripts: Post the list of outdated packages as a comment on PRs.

sw005320 commented:
Can you merge the update in the main repo to espnet3 and update this PR?

Masao-Someki commented:
Please merge #6192 first to merge master branch into espnet-3 and this PR

Masao-Someki commented:
@sw005320 
I've resolved the conflicts and all CI checks have passed. Could you please review and merge this PR if everything looks good to you?

Masao-Someki commented:
Hi @sw005320 or @Fhrozen, could you rerun the CI, please? It seems the CI fails when it tries to download a package via pip, and it should be passed.

sw005320 commented:
OK, I merged it.
What is the next?
